router/pkg/pubsub/nats.go

```diff
@@ -43,46 +57,87 @@ func (p *natsPubSub) ID() string {
 
 func (p *natsPubSub) ensureConn() error {
 	if p.conn == nil {
-		return errors.New("NATS is not configured")
+		return newEDFSNatsError(errors.New("NATS is not configured"))
 	}
 	return nil
 }
 
-func (p *natsPubSub) Subscribe(ctx context.Context, topic string, updater resolve.SubscriptionUpdater) error {
+func (p *natsPubSub) Subscribe(ctx context.Context, subjects []string, updater resolve.SubscriptionUpdater, streamConfiguration *pubsub_datasource.StreamConfiguration) error {
 	if err := p.ensureConn(); err != nil {
 		return err
 	}
-	sub, err := p.conn.SubscribeSync(topic)
-	if err != nil {
-		return fmt.Errorf("error subscribing to NATS topic %s: %w", topic, err)
-	}
-	go func() {
-		for {
-			msg, err := sub.NextMsgWithContext(ctx)
-			if err != nil {
-				_ = sub.Unsubscribe()
-				return
+	if streamConfiguration != nil {
+		js, err := jetstream.New(p.conn)
+		if err != nil {
+			return err
+		}
+
+		consumer, err := js.CreateOrUpdateConsumer(ctx, streamConfiguration.StreamName, jetstream.ConsumerConfig{
+			Durable:        streamConfiguration.Consumer, // Durable consumers are not removed automatically regardless of the InactiveThreshold
+			FilterSubjects: subjects,
+		})
+		if err != nil {
+			return newEDFSNatsError(fmt.Errorf(`failed to create or update consumer "%s": %w`, streamConfiguration.Consumer, err))
+		}
+		if consumer == nil {
+			return newEDFSNatsError(fmt.Errorf(`consumer "%s" is nil; it is likely the nats stream "%s" does not exist`, streamConfiguration.Consumer, streamConfiguration.StreamName))
+		}
+		go func() {
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+				}
+				msgBatch, err := consumer.FetchNoWait(1)
+				if err != nil {
+					return
+				}
+				for msg := range msgBatch.Messages() {
+					err = msg.Ack()
+					if err != nil {
+						return
+					}
+					updater.Update(msg.Data())
+				}
 			}
-			updater.Update(msg.Data)
+		}()
+		return nil
+	}
+
+	for _, subject := range subjects {
+		subscription, err := p.conn.SubscribeSync(subject)
```

Comment:
I don't think this is the right approach.
I'd use ChanSubscribe multiple times with the same channel as the "sink".
Then use one single goroutine to read on this channel and emit to updater.
Let me know if you'd like to discuss this.

Reply:
Changed.